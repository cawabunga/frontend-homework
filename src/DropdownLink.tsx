import { connect } from 'react-redux'
import { createStructuredSelector } from 'reselect'

import { Company, ReduxState } from './types'
import { getIsDropdownMenuVisible, getSelectedCompany } from './selectors'
import { toggleDropdownMenuVisibility } from './actions'

import { DropdownMenu } from './DropdownMenu'

type ReduxProps = {
  isDropdownMenuVisible: boolean
  selectedCompany: Company | undefined
}

type DispatchProps = {
  toggleDropdownMenuVisibility: () => void
}

export const DropdownLink = ({
  isDropdownMenuVisible,
  toggleDropdownMenuVisibility,
  selectedCompany,
}: ReduxProps & DispatchProps) => (
  <>
    <div
      role={'button'}
      aria-label={'Profile settings'}
      className="nav__link"
      onClick={toggleDropdownMenuVisibility}
      data-test-nav-link
    >
      <div className="nav__link-text-wrapper">
        <div className="nav__link-text">Elon Musk</div>

        <div className="nav__link-subtext">
          {selectedCompany && selectedCompany.name}
        </div>
      </div>

      <i className="material-icons-outlined nav__link-icon">settings</i>
    </div>

    <DropdownMenu
      open={isDropdownMenuVisible}
      requestClose={toggleDropdownMenuVisibility}
    />
  </>
)

export default connect(
  createStructuredSelector<ReduxState, ReduxProps>({
    isDropdownMenuVisible: getIsDropdownMenuVisible,
    selectedCompany: getSelectedCompany,
  }),
  { toggleDropdownMenuVisibility }
)(DropdownLink)
