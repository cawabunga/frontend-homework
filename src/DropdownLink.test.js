import { shallow } from 'enzyme'
import { getByText, render } from '@testing-library/react'
import { DropdownLink } from './DropdownLink'

describe('<DropdownLink />', () => {
  const toggleDropdownMenuVisibilityMock = jest.fn()

  const shallowRender = (props) =>
    shallow(
      <DropdownLink
        toggleDropdownMenuVisibility={toggleDropdownMenuVisibilityMock}
        selectedCompany={{ id: 1, name: 'Viljatootja AS' }}
        {...props}
      />
    )

  it('renders without dropdown menu', () => {
    expect(shallowRender({ isDropdownMenuVisible: false })).toMatchSnapshot()
  })

  it('renders with dropdown menu', () => {
    expect(shallowRender({ isDropdownMenuVisible: true })).toMatchSnapshot()
  })

  it('calls toggleDropdownMenuVisibility when nav link is clicked', () => {
    const component = shallowRender({ isDropdownMenuVisible: false })
    component.find('[data-test-nav-link]').simulate('click')

    expect(toggleDropdownMenuVisibilityMock).toHaveBeenCalled()
  })

  it('shows selected company name inside menu opener', () => {
    const { getByRole } = render(
      <DropdownLink
        isDropdownMenuVisible={false}
        selectedCompany={{ id: 1, name: 'Foo' }}
        toggleDropdownMenuVisibility={() => {}}
      />
    )
    const button = getByRole('button', { name: /profile/i })
    expect(() => getByText(button, 'Foo')).not.toThrow()
  })
})
