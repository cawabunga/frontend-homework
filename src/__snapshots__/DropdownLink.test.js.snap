// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<DropdownLink /> renders with dropdown menu 1`] = `
<Fragment>
  <div
    aria-label="Profile settings"
    className="nav__link"
    data-test-nav-link={true}
    onClick={[MockFunction]}
    role="button"
  >
    <div
      className="nav__link-text-wrapper"
    >
      <div
        className="nav__link-text"
      >
        Elon Musk
      </div>
      <div
        className="nav__link-subtext"
      >
        Viljatootja AS
      </div>
    </div>
    <i
      className="material-icons-outlined nav__link-icon"
    >
      settings
    </i>
  </div>
  <DropdownMenu />
</Fragment>
`;

exports[`<DropdownLink /> renders without dropdown menu 1`] = `
<Fragment>
  <div
    aria-label="Profile settings"
    className="nav__link"
    data-test-nav-link={true}
    onClick={[MockFunction]}
    role="button"
  >
    <div
      className="nav__link-text-wrapper"
    >
      <div
        className="nav__link-text"
      >
        Elon Musk
      </div>
      <div
        className="nav__link-subtext"
      >
        Viljatootja AS
      </div>
    </div>
    <i
      className="material-icons-outlined nav__link-icon"
    >
      settings
    </i>
  </div>
</Fragment>
`;
